pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticsearchClient.java

```diff
@@ -0,0 +1,565 @@
+                record.ack();
+                return true;
+            }
+            record.fail();
+            return false;
+        } catch (final Exception ex) {
+            log.debug("index failed id=" + id, ex);
+            record.fail();
+            throw ex;
+        }
+    }
+
+    /**
+     * Flushes the bulk processor.
+     */
+    public void flush() {
+        bulkProcessor.flush();
+    }
+
+    public void close() {
+        try {
+            if (bulkProcessor != null) {
+                bulkProcessor.awaitClose(5000L, TimeUnit.MILLISECONDS);
+            }
+        } catch (InterruptedException e) {
+            log.warn("Elasticsearch bulk processor close error:", e);
+        }
+        try {
+            this.executorService.shutdown();
+            if (this.client != null) {
+                this.client.close();
+            }
+        } catch (IOException e) {
+            log.warn("Elasticsearch client close error:", e);
+        }
+    }
+
+    private void checkNotFailed() throws Exception {
+        if (irrecoverableError.get() != null) {
+            throw irrecoverableError.get();
+        }
+    }
+
+    private void checkIndexExists(Optional<String> topicName) throws IOException {
+        String indexName = indexName(topicName);
+        if (!indexCache.contains(indexName)) {
+            synchronized (this) {
+                if (!indexCache.contains(indexName)) {
+                    createIndexIfNeeded(indexName);
+                    indexCache.add(indexName);
+                }
+            }
+        }
+    }
+
+    private String indexName(Optional<String> topicName) throws IOException {
+        if (config.getIndexName() != null) {
+            // Use the configured indexName if provided.
+            return config.getIndexName();
+        }
+        if (!topicName.isPresent()) {
+            throw new IOException("Elasticsearch index name configuration and topic name are empty");
+        }
+        return topicToIndexName(topicName.get());
+    }
+
+    @VisibleForTesting
+    public String topicToIndexName(String topicName) {
+        return topicToIndexCache.computeIfAbsent(topicName, k -> {
+            // see elasticsearch limitations https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-create-index.html#indices-create-api-path-params
+            String indexName = topicName.toLowerCase(Locale.ROOT);
+
+            // truncate to the max bytes length
+            while (indexName.getBytes(StandardCharsets.UTF_8).length > 255) {
+                indexName = indexName.substring(0, indexName.length() - 1);
+            }
+            if (indexName.length() <= 0 || !indexName.matches("[a-zA-Z\\.0-9][a-zA-Z_\\.\\-\\+0-9]*")) {
+                throw new RuntimeException(new IOException("Cannot convert the topic name='" + topicName + "' to a valid elasticsearch index name"));
+            }
+            log.debug("Translate topic={} to index={}", k, indexName);
```

Comment:
please use `if (log.isDebugEnabled()) `here and in other points

Reply:
Ok, fixed.